src/achievements/achievements_status.cpp

```diff
@@ -157,22 +374,198 @@ void AchievementsStatus::sync(const std::vector<uint32_t> & achieved_ids)
+    int max_kart_hits = -1;
+    if (type == UP_ACHIEVEMENT_DATA)
+    {
+        goal_string[0] = m_ach_enum_to_xml[(int)enum_id];
+    } // if type == UP_ACHIEVEMENT_DATA)
+    else if (type ==  UP_TRACK_DATA)
+    {
+        goal_string[0] = m_tr_enum_to_xml[(int)enum_id]; // The "one-track at least" goal
+        goal_string[1] = m_tr_enum_to_xml[(int)enum_id+(int)TR_DATA_NUM]; // The "all tracks" goal
+
+        for (unsigned int i=0;i<m_track_stats.size();i++)
+        {
+            // ignore addons tracks (compare returns 0 when the values are equal)
+            // Note: non-official tracks installed directly in the tracks folder
+            // are considered as officials by this method.
+            if (m_track_stats[i].ident.compare(0 /*start of sub-string*/,5/*length*/,"addon") == 0)
+                continue;
+
+            if (m_track_stats[i].track_data[enum_id] > max_across_tracks)
+                max_across_tracks = m_track_stats[i].track_data[enum_id];
+            if (m_track_stats[i].track_data[enum_id] < min_across_tracks)
+                min_across_tracks = m_track_stats[i].track_data[enum_id];
+        }
+    } // if type ==  UP_TRACK_DATA
+    else if (type == UP_KART_HITS)
+    {
+        goal_string[0] = "hit-same-kart-1race";
+
+        for (unsigned int i=0;i<m_kart_hits.size();i++)
+        {
+            if (m_kart_hits[i] > max_kart_hits)
+                max_kart_hits = m_kart_hits[i];
+        }
+    } // if type == UP_KART_HITS
+
+    // Now that we know what string to look for, call an Achievement function
+    // which will look throughout the progress goalTree to update it
+    std::map<uint32_t, Achievement*>::const_iterator i;
+    for(i=m_achievements.begin(); i!=m_achievements.end(); i++)
+    {
+        // Don't bother checking again already completed achievements
+        if (i->second->isAchieved())
+            continue;
+
+        if (type == UP_ACHIEVEMENT_DATA)
+        {
+            i->second->setGoalValue(goal_string[0],m_variables[enum_id].counter);
+        }
+        else if (type ==  UP_TRACK_DATA)
+        {
+            i->second->setGoalValue(goal_string[0],max_across_tracks);
+            i->second->setGoalValue(goal_string[1],min_across_tracks);
+        }
+        else if (type == UP_KART_HITS)
+        {
+            i->second->setGoalValue(goal_string[0],max_kart_hits);
+        }
+    }
+}
+
 // ----------------------------------------------------------------------------
-void AchievementsStatus::onRaceEnd()
+void AchievementsStatus::increaseDataVar(unsigned int achieve_data_id, int increase)
 {
-    //reset all values that need to be reset
-    std::map<uint32_t, Achievement *>::iterator iter;
-    for ( iter = m_achievements.begin(); iter != m_achievements.end(); ++iter ) {
-        iter->second->onRaceEnd();
+    if (achieve_data_id<ACHIEVE_DATA_NUM)
+    {
+        m_variables[achieve_data_id].counter += increase;
+        if (m_variables[achieve_data_id].counter > 10000000)
```

Comment:
what is this hardcoded value? Finding a way without harcoding would be better. If no possible, using instead a "const" with a little explanation would be better